add working with ds proxy draft

vamsiraju opened the pull request.

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+}
+```
+
+#### Script outline
````
```suggestion
#### Setup functions within WipeProxy.sol
```

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+bytes32 cup = bytes32(cupid);
+```
+
+#### Set all allowances
````
```suggestion
#### Setup `setAllowance` function
```

KentonPrescott replied:
Since `setAllowance( )` has no dependencies on other functions internal the contract, I think it should be placed before the `wipeWithDai( )` section.

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+}
+```
+
+#### Checks
````
```suggestion
##### Checks
```

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
```diff
@@ -0,0 +1,437 @@
+
+#### Checks
+
+Ensure at least some Dai debt is being wiped in the transaction using a require statement
```
```suggestion
Within the function body, ensure at least some Dai debt is being wiped in the transaction using a require statement
```

KentonPrescott replied:
Important to either make this clear in the beginning with a single line that says all next subsections will be in the body of this function and/or a combination with reduced subsection title sizes (as I've shown on `Checks` and `Initialize variables`)

vamsiraju replied:
Added the line to clarify that the contents of the next sections would go within the `wipeWithDai()` function

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+require(wad > 0);
+```
+
+#### Initialize variables
````
```suggestion
##### Initialize variables
```

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+```text
+tub.wipe(cup, wad);
+```
+
````
This could be a good point share a link to or a section of the entire `wipeWithDai( )` function or `WipeProxy.sol`

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+Add the `TubLike` interface to interact with CDPs on the Tub contract
+
+```text
+interface TubLike {
````
```suggestion
contract TubLike {
```

KentonPrescott replied:
In Solc ^0.5.4 (default dapp init), `interface` -> `contract`.. extent revisions to other interfaces

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+
+```text
+import "ds-math/math.sol";
+
````
Probably because I'm doing this late at night.. I initially thought to include all the interfaces in the `WipeProxy` Contract. This same mistake by others may be avoided if the interfaces are described right after `import "ds-math/math.sol";`

vamsiraju replied:
I split this into two separate steps and moved them around to reflect the order in which code is added to the contract.

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,437 @@
+Wipe debt of the CDP with DAI and pay the stability fee with MKR available on the DSProxy contract.
+
+```text
+tub.wipe(cup, wad);
````
I got this error: 
```
Kentons-Macbook:wipe-proxy kentonprescott$ dapp build
+ dapp clean
+ rm -rf out
+ solc --overwrite ds-math/=lib/ds-math/src/ ds-math=lib/ds-math/src/index.sol ds-test/=lib/ds-test/src/ ds-test=lib/ds-test/src/index.sol --abi --bin --bin-runtime --metadata /=/ -o out/ src/WipeProxy.t.sol
src/WipeProxy.sol:71:21: Compiler error: Stack too deep, try removing local variables.
      tub.wipe(cup, wad);
                    ^-^
Kentons-Macbook:wipe-proxy kentonprescott$ 
```

From this function construction: 
```
function wipeWithDai(
    address _tub,
    address _DAIExchange,
    address _MKRExchange,
    uint cupid,
    uint wad
  ) public {
      require(wad > 0);
      TubLike tub = TubLike(_tub);
      UniswapExchangeLike daiEx = UniswapExchangeLike(_DAIExchange);
      UniswapExchangeLike mkrEx = UniswapExchangeLike(_MKRExchange);

      TokenLike dai = tub.sai();
      TokenLike mkr = tub.gov();
      PepLike pep =   tub.pep();

      bytes32 cup = bytes32(cupid);

      setAllowance(dai, _tub);
      setAllowance(mkr, _tub);
      setAllowance(dai, _DAIExchange);

      (bytes32 val, bool ok) = pep.peek();

      uint daiFee = rmul(wad, rdiv(tub.rap(cup), tub.tab(cup)));
      uint mkrFee = wdiv(rmul(wad, rdiv(tub.rap(cup), tub.tab(cup))), uint(val));

      uint ethAmt = mkrEx.getEthToTokenOutputPrice(mkrFee);
      uint daiAmt = daiEx.getTokenToEthOutputPrice(ethAmt);

      daiAmt = add(wad, daiAmt);
      require(dai.transferFrom(msg.sender, address(this), daiAmt));

      if(ok && val != 0) {
          daiEx.tokenToTokenSwapOutput(mkrFee, daiAmt, uint(999000000000000000000), uint(1645118771), address(mkr));
      }

      tub.wipe(cup, wad);


  }
```

vamsiraju replied:
Good catch! I refactored code to remove `daiFee` but forgot to remove the corresponding step from the guide.

KentonPrescott commented:
Give me a shout when the Stack depth error is resolved, and I can review the rest of the guide 😄

vamsiraju commented:
Great suggestions Kenton! I added changes to the PR. The error was because of a redundant step, `daiFee` is not required and I removed it from the guide.

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
```diff
@@ -0,0 +1,488 @@
+
+You have to create a `~/.sethrc` file and configure it with these values to work with the Kovan testnet,
+
+* `SETH_CHAIN=kovan`
```
Spent like 30 min trying to understand why ~/.sethrc wasn't running/being checked... I realized I missed the `export` command before every line! haha!  Could we add these commands to every config line? Such as `export SETH_CHAIN=kovan`, `export ETH_FROM ... `, etc?

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,488 @@
+seth send 0x64a436ae831c1672ae81f674cab8b6775df3475c 'build()'
+```
+
+This transaction might fail if you already have deployed a DSProxy contract before from this address. You can check if you have one now with this command
````
Such a valuable piece of information! :)

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,488 @@
+seth call 0x64a436ae831c1672ae81f674cab8b6775df3475c 'proxies(address)(address)' 0xYourAddressHere
+```
+
+Make a note of the returned DSProxy contract address to use in a command later.
````
This would be a great point to set a variable that can be used in the call to `execute( )`
Ex: `Let's set the returned DSProxy contract address to be used later:
export MyProxy= 0xYourReturnedProxyAddress

KentonPrescott commented on devtools/working-with-dsproxy/working-with-dsproxy.md:
````diff
@@ -0,0 +1,488 @@
+* Calldata to execute the `wipeWithDai` script
+
+```bash
+seth send 0xd5282809a9d40311e06e0992b773075bbd40dde5 'execute(address,bytes memory)' $WIPEPROXY $CALLDATA
````
```suggestion
seth send $MYPROXY 'execute(address,bytes memory)' $WIPEPROXY $CALLDATA
```

KentonPrescott replied:
See above comment with respect to $MYPROXY

KentonPrescott replied:
Unfortunately, I'm failing at this line :(
https://kovan.etherscan.io/tx/0x90f386a0fdb74f4977d3149ce727c9fba612497f22247a9f711a6d50517bf3bf

vamsiraju commented:
@KentonPrescott I've added the missing token allowance step along with the other suggestions you had.
